src/librustc_typeck/check/mod.rs

```diff
@@ -4560,9 +4561,9 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     /// Report errors if the provided parameters are too few or too many.
     fn check_path_parameter_count(&self,
                                   span: Span,
-                                  segment: &mut Option<(&hir::PathSegment, &ty::Generics)>) {
+                                  segment: &mut Option<(&hir::PathSegment, Rc<ty::Generics>)>) {
```

Comment:
Maybe you shouldn't change anything about `ty::Generics`, if you can avoid it (since we're keeping at least some of the typed drop-ful arenas for now).